misk-crypto/src/main/kotlin/misk/crypto/CiphertextFormat.kt

```diff
@@ -0,0 +1,308 @@
+        throw EncryptionContextMismatchException("encryption context doesn't match")
+      }
+      val ciphertext = readCiphertext(src)
+      return Pair(ciphertext, aad)
+    }
+
+    /**
+     * Serializes the encryption context to a [ByteArray] so it could be passed to Tink's
+     * encryption/decryption methods.
+     */
+    fun serializeEncryptionContext(context: Map<String, String>?): ByteArray? {
+      if (context == null || context.isEmpty()) {
+        return null
+      }
+
+      val buff = ByteStreams.newDataOutput()
+      val contextSize = encodeVarInt(context.size)
+      var bytesWritten = 0
+      buff.write(contextSize)
+      bytesWritten +=  contextSize.size
+      context.toSortedMap(compareBy { it }).forEach { (k, v) ->
+        if (k.isEmpty() || v.isEmpty()) {
+          throw InvalidEncryptionContextException("empty key or value")
+        }
+        val key = k.toByteArray(Charsets.UTF_8)
+        val value = v.toByteArray(Charsets.UTF_8)
+        if (key.size >= Short.MAX_VALUE) {
+          throw InvalidEncryptionContextException("key is too long")
+        }
+        if (value.size >= Short.MAX_VALUE) {
+          throw InvalidEncryptionContextException("value is too long")
+        }
+        val keySize = encodeVarInt(key.size)
+        buff.write(keySize)
+        bytesWritten += keySize.size
+        buff.write(key)
+        bytesWritten += key.size
+        val valueSize = encodeVarInt(value.size)
+        buff.write(valueSize)
+        bytesWritten += valueSize.size
+        buff.write(value)
+        bytesWritten += value.size
+        if (bytesWritten >= Short.MAX_VALUE.toInt()) {
+          throw InvalidEncryptionContextException("encryption context is too long")
+        }
+      }
+      val aad = buff.toByteArray()
+      return aad
+    }
+
+    @VisibleForTesting
+    internal fun deserializeEncryptionContext(aad: ByteArray?): Map<String, String>? {
+      if (aad == null) {
+        return null
+      }
+      val src = DataInputStream(ByteArrayInputStream(aad))
+      val entries = decodeVarInt(src)
+      if (entries == 0) {
+        return null
+      }
+      return (1..entries).map {
+        val keySize = decodeVarInt(src)
+        val keyBytes = ByteArray(keySize)
+        src.readFully(keyBytes)
+        val valueSize = decodeVarInt(src)
+        val valueBytes = ByteArray(valueSize)
+        src.readFully(valueBytes)
+        keyBytes.toString(Charsets.UTF_8) to valueBytes.toString(Charsets.UTF_8)
+      }.toMap()
+    }
+
+    private fun readCiphertext(src: DataInputStream) : ByteArray {
+      val ciphertextStream = ByteArrayOutputStream()
+      var readByte = src.read()
+      while(readByte >= 0) {
+        ciphertextStream.write(readByte)
+        readByte = src.read()
+      }
+      return ciphertextStream.toByteArray()
+    }
+
+    private const val SEPTET = (1 shl 7) -1
+    private const val HAS_MORE_BIT = 1 shl 7
+
+    private fun encodeVarInt(integer: Int): ByteArray {
```

Comment:
You might wanna consider using an Okio buffer for all of this. It'll dynamically resize itself as necessary and has many optimizations to reduce allocations and keep the code fast.

One recent example:
https://publicobject.com/2020/07/26/optimizing-new-byte/